common/types.go

```diff
@@ -371,20 +660,146 @@ func (a *Address) Scan(src interface{}) error {
 }
 
 // Value implements valuer for database/sql.
-func (a Address) Value() (driver.Value, error) {
+func (a InternalAddress) Value() (driver.Value, error) {
 	return a[:], nil
 }
 
 // IsInChainScope checks if an address is a valid account in our node's sharded address space
-func (a Address) IsInChainScope() bool {
-	if a == ZeroAddr {
+func (a InternalAddress) IsInChainScope() bool { // Eventually this will just return true
+	if a.Hash() == ZeroAddr.Hash() {
 		return true
 	}
-	return NodeLocation.ContainsAddress(a)
+	return NodeLocation.ContainsAddress(Address{&a})
+}
+
+func (a InternalAddress) Location() *Location {
+	return &NodeLocation
+}
+
+// Bytes gets the string representation of the underlying address.
+func (a ExternalAddress) Bytes() []byte { return a[:] }
+
+// Hash converts an address to a hash by left-padding it with zeros.
+func (a ExternalAddress) Hash() Hash { return BytesToHash(a[:]) }
+
+// Hex returns an EIP55-compliant hex string representation of the address.
+func (a ExternalAddress) Hex() string {
+	return string(a.checksumHex())
+}
+
+// String implements fmt.Stringer.
+func (a ExternalAddress) String() string {
+	return a.Hex()
+}
+
+func (a *ExternalAddress) checksumHex() []byte {
+	buf := a.hex()
+
+	// compute checksum
+	sha := sha3.NewLegacyKeccak256()
+	sha.Write(buf[2:])
+	hash := sha.Sum(nil)
+	for i := 2; i < len(buf); i++ {
+		hashByte := hash[(i-2)/2]
+		if i%2 == 0 {
+			hashByte = hashByte >> 4
+		} else {
+			hashByte &= 0xf
+		}
+		if buf[i] > '9' && hashByte > 7 {
+			buf[i] -= 32
+		}
+	}
+	return buf[:]
+}
+
+func (a ExternalAddress) hex() []byte {
+	var buf [len(a)*2 + 2]byte
+	copy(buf[:2], "0x")
+	hex.Encode(buf[2:], a[:])
+	return buf[:]
+}
+
+// Format implements fmt.Formatter.
+// Address supports the %v, %s, %v, %x, %X and %d format verbs.
+func (a ExternalAddress) Format(s fmt.State, c rune) {
+	switch c {
+	case 'v', 's':
+		s.Write(a.checksumHex())
+	case 'q':
+		q := []byte{'"'}
+		s.Write(q)
+		s.Write(a.checksumHex())
+		s.Write(q)
+	case 'x', 'X':
+		// %x disables the checksum.
+		hex := a.hex()
+		if !s.Flag('#') {
+			hex = hex[2:]
+		}
+		if c == 'X' {
+			hex = bytes.ToUpper(hex)
+		}
+		s.Write(hex)
+	case 'd':
+		fmt.Fprint(s, ([len(a)]byte)(a))
+	default:
+		fmt.Fprintf(s, "%%!%c(address=%x)", c, a)
+	}
+}
+
+// SetBytes sets the address to the value of b.
+// If b is larger than len(a), b will be cropped from the left.
+func (a *ExternalAddress) SetBytes(b []byte) {
```

Comment:
Same comment as above, for InternalAddress method

Reply:
I agree that `SetBytes()` does not need to exist for the `Address` type, but the `Address` constructor uses `SetBytes` on the inner AddressData type, so it's still required for InternalAddress and ExternalAddress

Reply:
However I agree that it is dangerous to allow other functions to call `SetBytes()` on an `InternalAddress` because it doesn't do any scope checks so I've made it a private function